metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala

```diff
@@ -2215,15 +2215,21 @@ class MetalsLanguageServer(
   }
 
   private def indexWorkspaceSources(): Unit = {
+    case class SourceToIndex(
+        source: AbsolutePath,
+        sourceItem: AbsolutePath,
+        targets: Iterable[b.BuildTargetIdentifier]
+    )
+    val sourcesToIndex = mutable.ListBuffer.empty[SourceToIndex]
     for {
       (sourceItem, targets) <- buildTargets.sourceItemsToBuildTargets
       source <- sourceItem.listRecursive
       if source.isScalaOrJava
-    } {
-      targets.asScala.foreach { target =>
-        buildTargets.linkSourceFile(target, source)
-      }
-      indexSourceFile(source, Some(sourceItem), targets.asScala.headOption)
+    } { sourcesToIndex += SourceToIndex(source, sourceItem, targets.asScala) }
+
+    sourcesToIndex.par.foreach { f =>
+      f.targets.foreach(buildTargets.linkSourceFile(_, f.source))
```

Comment:
If `buildTargets.linkSourceFile` updates `ConcurrentHashSet` then doing it in parallel will have any effect? Threads will be blocked and there won't be any benefits from `par`.

I wonder what results will look like if we extract `f.targets.foreach(buildTargets.linkSourceFile(_, f.source))` to separate `foreach`.

When you mentioned Kafka, did you refer to the original repo? I want to do some experiments on my own.

Reply:
Good point - I should probably move that back to the first for loop.
I'm using kafka from [here](https://github.com/apache/kafka).  The new "Start Metals" button is very handy for repeatedly checking out indexing speed 😄

Reply:
Got the results from my machine (MBP 2019).
before -> from this branch -> after moving back `buildTargets.linkSourceFile` to the first loop
Kafka: 25s -> 13s -> 12s
Metals: 15s -> 12s -> 11s
Metals (second attempt): 12s -> 8s -> 8s

Reply:
Looks like I need a new machine